javascripts/discourse/services/docs-sidebar.js

```diff
@@ -0,0 +1,144 @@
+import { MAIN_PANEL } from "discourse/lib/sidebar/panels";
+import { parseSidebarStructure } from "../lib/cooked-html-parser";
+
+export const SIDEBAR_DOCS_PANEL = "discourse_sidebar_docs";
+
+export default class DocsSidebarService extends Service {
+  @service appEvents;
+  @service router;
+  @service sidebarState;
+  @service store;
+
+  #contentCache = new Map();
+
+  @tracked _activeTopicId;
+  @tracked _currentSectionsConfig = null;
+  @tracked _loading = false;
+
+  constructor() {
+    super(...arguments);
+
+    this.appEvents.on("page:changed", this, this.#maybeForceDocsSidebar);
+  }
+
+  get activeCategory() {
+    return (
+      this.router.currentRoute?.attributes?.category ||
+      this.router.currentRoute?.parent?.attributes?.category
+    );
+  }
+
+  get allSectionsExpanded() {
+    return this.sectionsConfig?.every((sectionConfig) => {
+      return !this.sidebarState.collapsedSections.has(
+        `sidebar-section-${sectionConfig.name}-collapsed`
+      );
+    });
+  }
+
+  get isEnabled() {
+    return !!this._activeTopicId;
+  }
+
+  get isVisible() {
+    return this.sidebarState.isCurrentPanel(SIDEBAR_DOCS_PANEL);
+  }
+
+  get loading() {
+    return this._loading;
+  }
+
+  get sectionsConfig() {
+    return this._currentSectionsConfig || [];
+  }
+
+  hideDocsSidebar() {
+    this.sidebarState.setPanel(MAIN_PANEL);
+  }
+
+  showDocsSidebar() {
+    this.sidebarState.setPanel(SIDEBAR_DOCS_PANEL);
+    this.sidebarState.setSeparatedMode();
+    this.sidebarState.hideSwitchPanelButtons();
+  }
+
+  toggleSidebarPanel() {
+    if (this.isVisible) {
+      this.hideDocsSidebar();
+    } else {
+      this.showDocsSidebar();
+    }
+  }
+
+  #findSettingsForActiveCategory() {
+    let category = this.activeCategory;
+
+    while (category != null) {
+      const matchingSetting = settings.categories_toc_topics.find(
+        // eslint-disable-next-line no-loop-func
+        (setting) => setting.category[0] === category.id
+      );
+
+      if (matchingSetting) {
+        return matchingSetting;
+      }
+
+      category = category.parentCategory;
+    }
+  }
+
+  #maybeForceDocsSidebar() {
+    const newActiveTopicId = this.#findSettingsForActiveCategory()?.topic_id;
+
+    if (this._activeTopicId !== newActiveTopicId) {
+      this.#setSidebarContent(newActiveTopicId);
+    }
+  }
+
+  async #setSidebarContent(topic_id) {
+    this._activeTopicId = topic_id;
+
+    if (!this._activeTopicId) {
+      this.hideDocsSidebar();
+      return;
+    }
+
+    this._currentSectionsConfig = this.#contentCache.get(this._activeTopicId);
+
+    if (this._currentSectionsConfig) {
+      this.showDocsSidebar();
+      return;
+    }
+
+    await this.#fetchTopicContent(topic_id);
+  }
+
+  async #fetchTopicContent(topic_id) {
+    this._loading = true;
+    this.showDocsSidebar();
+
+    try {
+      // leverages the post stream API to fetch only the first post
+      const data = await ajax(`/t/${topic_id}/posts.json`, {
```

Comment:
I'm not sure but you might be able to use the `store` service to find the post or topic if it's already loaded in memory from visiting the topic page:

https://github.com/discourse/discourse/blob/8eee592f6a286c06f1400f6a5f7c93b5d2407cea/app/assets/javascripts/discourse/app/services/composer.js#L863-L863

Reply:
The `post` store doesn't work because we need the `post_id`.

I tried the `topic` store before using the request directly, but it brings too much extra stuff in the request, and up to 20 posts.

I believe this is a good compromise, the TC caches the data and, at least on Meta, these topics are invisible, so the chance of having a hot cache in the store is very, very low.

Reply:
Okay cool, I wasn't sure whether this would actually work, go with what works! :+1: